cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/DefaultApplicationUpdateContext.java

```diff
@@ -0,0 +1,164 @@
+    this.applicationArtifactId = applicationArtifactId;
+    this.configString = configString;
+    this.updateActions = updateActions;
+  }
+
+  @Override
+  public List<ApplicationConfigUpdateAction> getUpdateActions() {
+    return updateActions;
+  }
+
+  @Override
+  public <C extends Config> C getConfig(Type configType) {
+    C appConfig;
+
+    if (configString.isEmpty()) {
+      try {
+        appConfig = ((Class<C>) configType).newInstance();
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Issue in creating config class of type " + configType.getTypeName(), e);
+      }
+    } else {
+      try {
+        appConfig = GSON.fromJson(configString, configType);
+      } catch (JsonSyntaxException e) {
+        throw new IllegalArgumentException("Invalid JSON application configuration was provided. Please check the"
+                                           + " syntax.", e);
+      }
+    }
+    return appConfig;
+  }
+
+  @Override
+  public String getConfigAsString() {
+    return configString;
+  }
+
+  // Returns plugin artifacts using given filters in ascending order.
+  // TODO: Pass ArtifactSortOrder as argument for better flexibility.
+  @Override
+  public List<ArtifactId> getPluginArtifacts(String pluginType, String pluginName, ArtifactScope pluginScope,
+                                             @Nullable ArtifactVersionRange pluginRange, int limit) {
+    List<ArtifactId> pluginArtifacts = new ArrayList<>();
+    NamespaceId pluginArtifactNamespace = ArtifactScope.SYSTEM.equals(pluginScope) ? NamespaceId.SYSTEM : namespaceId;
+
+    Predicate<io.cdap.cdap.proto.id.ArtifactId> predicate = new Predicate<io.cdap.cdap.proto.id.ArtifactId>() {
+      @Override
+      public boolean apply(io.cdap.cdap.proto.id.ArtifactId input) {
+        // should check if the artifact is from SYSTEM namespace, if not, check if it is from the scoped namespace.
+        // should check if plugin is in given range if provided.
+        return (((pluginScope == null && NamespaceId.SYSTEM.equals(input.getParent()))
+               || pluginArtifactNamespace.equals(input.getParent())) &&
+               (pluginRange == null || pluginRange.versionIsInRange(new ArtifactVersion(input.getVersion()))));
+      }
+    };
+
+    try {
+      Map<ArtifactDescriptor, PluginClass> plugins =
+          artifactRepository.getPlugins(pluginArtifactNamespace,
+                                        Artifact.from(Namespace.fromEntityId(namespaceId), applicationArtifactId),
+                                        pluginType, pluginName, predicate, limit, ArtifactSortOrder.ASC);
+      for (Map.Entry<ArtifactDescriptor, PluginClass> pluginsEntry : plugins.entrySet()) {
+        ArtifactId plugin = pluginsEntry.getKey().getArtifactId();
+        // Only consider non-SNAPSHOT plugins for upgrade.
+        // TODO: Consider making this check optional. Helpful for integration tests.
+        if (!plugin.getVersion().isSnapshot()) {
```

Comment:
It is better not to have this logic in this class, but rather an argument to this class (e.g. `boolean allowSnapshot`)